File: web3py/Phase3Bot.py
from web3 import Web3
import config
import math
import random
import time


w3 = Web3(Web3.HTTPProvider(config.infuraProvider))
smartContractAddress = Web3.toChecksumAddress(config.smartContractAddress)


print(w3.isConnected())

casinoContract = w3.eth.contract(address=smartContractAddress, abi=config.abi)
accounts = w3.eth.account.privateKeyToAccount(config.casinoPrivateKey)
account = accounts.address
#account = w3.eth.accounts[0]
# casino account


# Slight difference in web3js and web3py. Instead of .methods, we use .functions
# print(casinoContract.functions.maxBet().call())



def getPhase3Games():
    activeGames = []
    temp = casinoContract.functions.getPhase3Games().call({'from': account})
    for i in range(len(temp)):
        if (temp[i] != '0x0000000000000000000000000000000000000000'):
            activeGames.append(temp[i])
    return activeGames


def phase3Response(activeGames):

    if len(activeGames) > 0:
        print("Detected the following games")
        print(activeGames)
    for i in activeGames:
        activeGame = i
        print("Distribute cards for user and dealer for address:", activeGame)

        nonce = w3.eth.getTransactionCount(account)
        transaction = casinoContract.functions.distribute(activeGame).buildTransaction(
            {'nonce': nonce, 'from': account, 'to': activeGame})
        signed_tx = w3.eth.account.sign_transaction(
            transaction, config.casinoPrivateKey)
        response = w3.eth.send_raw_transaction(signed_tx.rawTransaction)
        print(response)
        print("Done")


while True:
    print("Scanning games...")
    print()
    phase3Response(getPhase3Games())

    time.sleep(5)

File: BackEnd/Phase3Bot.js
//IMPORTS
const config = require("./config")


//GLOBAL VARS
const Web3 = require("web3");
const provider = new Web3.providers.HttpProvider(config.web3Provider);
const web3 = new Web3(provider);
web3.eth.net.isListening()
.then(() => console.log('web3 is connected'))
.catch(e => console.log('Wow. Something went wrong'));


const casinoContract = new web3.eth.Contract(config.abi,config.smartContractAddress)


web3.eth.accounts.wallet.add('0x' + config.casinoPrivateKey )


var accounts = web3.eth.accounts.wallet
var account = accounts["0"]
var myAddress = account["address"]

// getActiveGames -> Retrieves all of the games that are in Phase 2



const getPhase3Games = async () => {
   const temp = await casinoContract.methods.getPhase3Games().call({from: myAddress})
   activeGames = []
   for (i = 0; i < temp.length; i++){
       if (temp[i] != '0x0000000000000000000000000000000000000000'){
           activeGames.push(temp[i])
       }
   }
   return activeGames
}


const phase3Response =  async (activeGames, nonce) => {
    if (activeGames.length > 0){
        console.log("Games in phase 3 detected:")
        console.log(activeGames)
        console.log()
    }
    for (i = 0; i < activeGames.length; i++){
        activeGame = activeGames[i]
        console.log("Distribute cards for user and dealer for address:",activeGame)


    
        const gasPrice = await web3.eth.getGasPrice();
        const gasEstimate = await casinoContract.methods.distribute(activeGame).estimateGas({ from: myAddress, to: activeGame  });
        
        const response = await casinoContract.methods.distribute(activeGame).send({from: myAddress, to: activeGame, gasPrice: gasPrice, gas: gasEstimate })
        console.log("Done!")
        console.log(response)
        console.log()
    }
}


const main = async () => {



     let nonce = await web3.eth.getTransactionCount(myAddress, 'pending')
    const phase3Bot =  async () => {
        console.log("- - - - - - - - - -")
        console.log("Scanning phase 3 games...")
        console.log()
        var activeGames =  await getPhase3Games()
        console.log(activeGames)
        nonce += 33
        await phase3Response(activeGames, nonce)
        setTimeout(phase3Bot,5000)
    }
    phase3Bot()


    
}

main()